modules/data-entry/src/main/frontend/src/questionnaireEditor/ReferenceInput.jsx

```diff
@@ -0,0 +1,292 @@
+      if (thisQuestionnaire["jcr:primaryType"] != "cards:Questionnaire") {
+        continue;
+      }
+
+      newFilterableFields.push([title, ...parseSectionOrQuestionnaire(thisQuestionnaire)]);
+    }
+
+    // We also need a filter over the subject
+    setOptions(newFilterableFields);
+    setTitleMap(newFilterableTitles);
+    setPathMap(newFilterablePaths);
+  }
+
+  let parseSubjectTypeData = (subjectTypeJson) => {
+    // Parse through, but keep a custom field for the subject
+    let fields = [""];
+    let titles = {"": "None"};
+    let paths = {"": ""};
+
+    for (let subjectType of subjectTypeJson["rows"]) {
+      // For each reference, store its UUID and title
+      fields.push(subjectType["jcr:uuid"]);
+      titles[subjectType["jcr:uuid"]] = subjectType["label"];
+      paths[subjectType["jcr:uuid"]] = subjectType["@path"];
+    }
+    setOptions(fields);
+    setTitleMap(titles);
+    setPathMap(paths);
+  }
+
+  // From an array of fields, turn it into a react component
+  let GetReactComponentFromFields = (fields, nested=false) => {
+    return fields.map((path) => {
+      if (typeof path == "string") {
+        // Straight strings are MenuItems
+        // If we have a restriction, we might return nothing
+        if (restrictions && restrictions.length > 0 && !restrictions.includes(path)) {
+          return null;
+        }
+        return <MenuItem value={path} key={path} className={classes.categoryOption + (nested ? " " + classes.nestedSelectOption : "")}>{titleMap[path]}</MenuItem>
+      } else if (Array.isArray(path)) {
+        // Arrays represent Questionnaires of Sections
+        // which we'll need to turn into opt groups
+        return [<MenuItem className={classes.categoryHeader} disabled>{path[0]}</MenuItem>,
+          GetReactComponentFromFields(path.slice(1), true)];
+      }
+    })
+  }
+
+  let getRestrictions = (restrictingField) => {
+    let field = fieldsReader[restrictingField];
+    if (Array.isArray(field)) {
+      field = field[0];
+    }
+
+    if (field == undefined) {
+      setRestrictions(undefined);
+      return;
+    }
+
+    // Parse out the questionnaire name from the question path
+    // If this was selected from elsewhere in the form, we'll be given a path
+    let questionnaireMatch = field.match(/(\/Questionnaires\/.+?)\//);
+    let fetchRequest = null;
+    if (questionnaireMatch) {
+      let url = new URL(questionnaireMatch[1] + ".json", window.location.origin);
+      fetchRequest = fetchWithReLogin(globalLoginDisplay, url);
+    } else {
+      // If this is an existing value, we will be given a jcr:uuid instead
+      let url = new URL(`query?query=SELECT * FROM [nt:base] AS n WHERE n.'jcr:uuid'='${field}'`, window.location.origin);
+      fetchRequest = fetchWithReLogin(globalLoginDisplay, url)
+        .then((response) => response.ok ? response.json() : Promise.reject(response))
+        .then((json) => {
+          let nodePath = json["rows"][0]["@path"];
```

Comment:
For safety, let's do this instead:

```suggestion
          let nodePath = json["rows"]?.[0]?.["@path"];
          nodePath || Promise.reject("Invalid reference: " + field);
```